File: app/src/main/java/com/ericho/coupleshare/mvp/data/LoginRepository.java
package com.ericho.coupleshare.mvp.data;

import com.google.common.base.Preconditions;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Created by steve_000 on 12/6/2017.
 * for project CoupleShare
 * package name com.ericho.coupleshare.mvp.data
 */

public class LoginRepository implements LoginDataSource {

    private static LoginRepository INSTANCE ;

    private LoginDataSource loginDataSource;

    private LoginRepository(LoginDataSource dataSource){
        checkNotNull(dataSource);
        this.loginDataSource = dataSource;
    }

    public static LoginRepository getInstance(LoginDataSource dataSource){
        if(INSTANCE == null){
            INSTANCE = new LoginRepository(dataSource);
        }
        return INSTANCE;
    }

    public static void destroyInstance(){

        INSTANCE = null;
    }

    @Override
    public void login(String username, String password, LoginCallback callback) {
        loginDataSource.login(username, password, callback);
    }

    @Override
    public void register(String username, String password, RegisterCallback callback) {
        loginDataSource.register(username, password, callback);
    }
}

File: app/src/main/java/com/ericho/coupleshare/mvp/data/LoginRepository.kt
package com.ericho.coupleshare.mvp.data

import android.content.Context
import android.content.SharedPreferences
import com.ericho.coupleshare.App
import com.ericho.coupleshare.contant.Key

/**
 * Created by steve_000 on 7/7/2017.
 * for project CoupleShare
 * package name com.ericho.coupleshare.mvp.data
 */
class LoginRepository private constructor(private var dataSource: LoginDataSource):LoginDataSource {
    override fun login(username: String, password: String, callback: LoginDataSource.LoginCallback) {
        dataSource.login(username, password, object :LoginDataSource.LoginCallback{
            override fun onLoginSuccess() {
                logLoginAction(App.context,username)
                callback.onLoginSuccess()
            }

            override fun onLoginFailure(t: Throwable) {
                callback.onLoginFailure(t)
            }
        })
    }

    override fun register(username: String, password: String, callback: LoginDataSource.RegisterCallback) {
        dataSource.register(username, password, callback)
    }

    companion object {
        private var INSTANCE : LoginRepository? = null


        @JvmStatic
        fun getInstance(dataSource: LoginDataSource):LoginRepository{
            if (INSTANCE == null) {
                INSTANCE = LoginRepository(dataSource)
            }
            return INSTANCE as LoginRepository
        }
        @JvmStatic
        fun destroyInstance() {
            INSTANCE = null
        }
    }

    fun isLogin(context: Context):Boolean{
        val pref =  getPref(context)
        val res = pref?.getBoolean(Key.login,false) ?:false
        return res
    }
    fun getLoginName(context: Context):String{
        val pref =  getPref(context)
        val res = pref?.getString(Key.loginName,"") ?:""
        return res
    }

    fun logLoginAction(context: Context?,name:String){
        val pref =  getPref(context)
        pref?.edit()
                ?.putBoolean(Key.login,true)
                ?.putString(Key.loginName,name)
                ?.apply()
    }
    fun logout(context: Context){
        val pref = getPref(context)
        pref?.edit()
                ?.remove(Key.login)
                ?.remove(Key.loginName)
                ?.apply()
    }

    fun getPref(context: Context?):SharedPreferences?{
        return context?.getSharedPreferences(Key.pref_name,0)
    }
}